crates/madara/client/settlement_client/src/starknet/event.rs

```diff
@@ -0,0 +1,377 @@
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::sleep;
+
+type FutureType = Pin<Box<dyn Future<Output = anyhow::Result<(Option<EmittedEvent>, EventFilter)>> + Send>>;
+
+pub struct StarknetEventStream {
+    provider: Arc<JsonRpcClient<HttpTransport>>,
+    filter: EventFilter,
+    processed_events: HashSet<Felt>,
+    future: Option<FutureType>,
+    polling_interval: Duration,
+}
+
+impl StarknetEventStream {
+    pub fn new(provider: Arc<JsonRpcClient<HttpTransport>>, filter: EventFilter, polling_interval: Duration) -> Self {
+        Self { provider, filter, processed_events: HashSet::new(), future: None, polling_interval }
+    }
+}
+
+impl Stream for StarknetEventStream {
+    type Item = Option<anyhow::Result<CommonMessagingEventData>>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        if self.future.is_none() {
+            let provider = self.provider.clone();
+            let filter = self.filter.clone();
+            let processed_events = self.processed_events.clone();
+            let polling_interval = self.polling_interval;
+
+            async fn fetch_events(
+                provider: Arc<JsonRpcClient<HttpTransport>>,
+                mut filter: EventFilter,
+                processed_events: HashSet<Felt>,
+                polling_interval: Duration,
+            ) -> anyhow::Result<(Option<EmittedEvent>, EventFilter)> {
+                // Adding sleep to introduce delay
+                sleep(polling_interval).await;
+
+                let mut event_vec = Vec::new();
+                let mut page_indicator = false;
+                let mut continuation_token: Option<String> = None;
+
+                while !page_indicator {
+                    let events = provider
+                        .get_events(
+                            EventFilter {
+                                from_block: filter.from_block,
+                                to_block: filter.to_block,
+                                address: filter.address,
+                                keys: filter.keys.clone(),
+                            },
+                            continuation_token.clone(),
+                            1000,
+                        )
+                        .await?;
+
+                    event_vec.extend(events.events);
+                    if let Some(token) = events.continuation_token {
+                        continuation_token = Some(token);
+                    } else {
+                        page_indicator = true;
+                    }
+                }
+
+                let latest_block = provider.block_number().await?;
+
+                for event in event_vec.clone() {
+                    let event_nonce = event.data[1];
+                    if !processed_events.contains(&event_nonce) {
+                        return Ok((Some(event), filter));
+                    }
+                }
```

Comment:
```suggestion
        for event in event_vec.iter() {
            if let Some(nonce) = event.data.get(1) {
                if !processed_events.contains(nonce) {
                    processed_events.insert(*nonce);
                    return Ok((Some(event.clone()), filter));
                }
            }
        }
```
- Direct indexing is risky as it may cause a panic if the index is out of bounds
- Instead of cloning the entire event_vec, we should consider only extracting the necessary fields and clone only the matching event